src/Psy/ExecutionLoop/Loop.php

```diff
@@ -50,67 +91,81 @@ public function run(Shell $shell)
-                    $__psysh__->writeReturnValue($_);
-                } catch (BreakException $_e) {
-                    restore_error_handler();
-                    if (ob_get_level() > 0) {
-                        ob_end_clean();
-                    }
-                    $__psysh__->writeException($_e);
-
-                    return;
-                } catch (\Exception $_e) {
-                    restore_error_handler();
-                    if (ob_get_level() > 0) {
-                        ob_end_clean();
-                    }
-                    $__psysh__->writeException($_e);
-                }
+                $executionClosure($__psysh__);
 
-                // a bit of housekeeping
-                unset($__psysh_out__);
                 $__psysh__->afterLoop();
             } while (true);
         };
+    }
 
+    /**
+     * @param  Shell $shell
+     * @param  $code
+     * @return mixed
+     */
+    public function execute(Shell &$shell, $code)
+    {
+        $code = $this->cleaner->clean(array($code));
+
+        $executionClosure = $this->getExecutionClosure();
+        if (self::bindLoop()) {
+            $executionClosure = $this->setClosureShellScope($shell, $executionClosure);
+        }
+
+        return $executionClosure($shell, $code);
+    }
+
+    /**
+     * Run the execution loop.
+     *
+     * @param Shell $shell
+     */
+    public function run(Shell $shell)
+    {
+        $executionClosure = $this->getExecutionClosure();
         // bind the closure to $this from the shell scope variables...
         if (self::bindLoop()) {
-            $that = null;
-            try {
-                $that = $shell->getScopeVariable('this');
-            } catch (\InvalidArgumentException $e) {
-                // well, it was worth a shot
-            }
+            $executionClosure = $this->setClosureShellScope($shell, $executionClosure);
+        }
 
-            if (is_object($that)) {
-                $loop = $loop->bindTo($that, get_class($that));
-            } else {
-                $loop = $loop->bindTo(null, null);
+        $loop = $this->getLoopClosure($executionClosure);
+
+        try {
+            $loop($shell);
+        } catch (BreakException $_e) {
+            restore_error_handler();
+            if (ob_get_level() > 0) {
+                ob_end_clean();
             }
+            $shell->writeException($_e);
+
+            return;
+        }
+    }
+
+    /**
+     * @param $loop
+     * @return mixed
+     */
+    protected function setClosureShellScope(Shell $shell, $loop)
+    {
+        $that = null;
+        try {
+            $that = $shell->getScopeVariable('this');
+        } catch (\InvalidArgumentException $e) {
+            // well, it was worth a shot
+        }
+        if (is_object($that)) {
+            return $loop->bindTo($that, get_class($that));
         }
 
-        $loop($shell);
+        return $loop->bindTo(null);
```

Comment:
do we have a test for that, or can you explain me how to add it to the tests? thanks